climada/util/coordinates.py

```diff
@@ -459,6 +460,136 @@ def get_gridcellarea(lat, resolution=0.5, unit="ha"):
+
+    Example:
+    --------
+    >>> res = 10  #10° resolution
+    >>> area = compute_grid_cell_area(res = res)
+    >>> print(area.shape) # (180/10, 360/10)
+    (18, 36)
+    """
+
+    lat_bins = np.linspace(-90, 90, int(180 / res))  # lat bins
+    lon_bins = np.linspace(-180, 180, int(360 / res))
+
+    r_earth = 6371  # Earth's radius [km]
+    # Convert to radians
+    lat_bin_edges = np.radians(lat_bins)
+    lon_res_rad = np.radians(res)
+
+    # Compute area
+    areas = (
+        r_earth**2
+        * lon_res_rad
+        * np.abs(np.sin(lat_bin_edges[1:]) - np.sin(lat_bin_edges[:-1]))
+    )
+    # Broadcast to create a full 2D grid of areas
+    grid_area = np.tile(
+        areas[:, np.newaxis], (1, len(lon_bins) - 1)
+    )  # each row as same area
+
+    return grid_area, lat_bins, lon_bins
+
+
+def compute_grid_cell_area(
+    res: float = None,
+    bounds: tuple = None,
+    projection: str = "WGS84",
+    units: str = "km^2",
+) -> np.ndarray:
+    """
+    Compute the area of each grid cell in a latitude-longitude grid.
+
+    Parameters:
+    -----------
+    res: float
+        Grid resolution in degrees
+    bounds: tuple, dafault: None
+        (lon_min, lat_min, lon_max, lat_max) latitude and longitude bounds to compute grid cell area
+    projection: str
+        Ellipsoid or spherical projection to approximate Earth. To get the complete list of
+        projections call :py:meth:`pyproj.get_ellps_map()`. Widely used projections:
+        - "WGS84": Uses the WGS84 ellipsoid (default)
+        - "GRS80"
+        - "IAU76"
+        - "sphere": Uses a perfect sphere with Earth's mean radius (6371 km)
+    units: str (optional) Default "km^2"
+        units of the area. Either km^2 or m^2.
+
+    Returns:
+    --------
+    area: np.ndarray
+        A 2D numpy array of grid cell areas in km²
+    Example:
+    --------
+    >>> area = compute_grid_areas(res = 1, projection ="sphere", units = "m^2")
+    """
+    geod = Geod(ellps=projection)
+
+    if not bounds:
+        lon_min, lat_min, lon_max, lat_max = -180, -90, 180, 90
+    else:
+        lon_min, lat_min, lon_max, lat_max = bounds[0], bounds[1], bounds[2], bounds[3]
+
+    lat_edges: np.ndarray = np.linspace(lat_min, lat_max, int(180 / res))
+    lon_edges: np.ndarray = np.linspace(lon_min, lon_max, int(360 / res))
+
+    area = np.zeros((len(lat_edges) - 1, len(lon_edges) - 1))
+
+    # Iterate over consecutive latitude and longitude edges
+    for i, (lat1, lat2) in enumerate(zip(lat_edges[:-1], lat_edges[1:])):
+        for j, (lon1, lon2) in enumerate(zip(lon_edges[:-1], lon_edges[1:])):
+
+            # 5th point to close the loop
+            poly_lons = [lon1, lon2, lon2, lon1, lon1]
+            poly_lats = [lat1, lat1, lat2, lat2, lat1]
+
+            # Compute the area of the grid cell
+            poly_area, _ = geod.polygon_area_perimeter(poly_lons, poly_lats)
+            area[i, j] = abs(poly_area)  # Convert from m² to km²
```

Comment:
This looks quite inefficient. In general, nested loops in Python are not a good sign as these are often not vectorized.